cli/internal/run/real_run.go

```diff
@@ -94,18 +94,24 @@ func RealRun(
 	taskSummaries := []*runsummary.TaskSummary{}
 	execFunc := func(ctx gocontext.Context, packageTask *nodes.PackageTask, taskSummary *runsummary.TaskSummary) error {
 		deps := engine.TaskGraph.DownEdges(packageTask.TaskID)
-		mu.Lock()
-		taskSummaries = append(taskSummaries, taskSummary)
-		// don't hold the lock while we run ec.exec
-		mu.Unlock()
-
 		// deps here are passed in to calculate the task hash
 		taskExecutionSummary, err := ec.exec(ctx, packageTask, deps)
 		if err != nil {
 			return err
 		}
-		taskSummary.Execution = taskExecutionSummary
-		taskSummary.ExpandedOutputs = taskHashTracker.GetExpandedOutputs(taskSummary.TaskID)
+
+		// taskExecutionSummary will be nil if the task never executed
+		// (i.e. if the workspace didn't implement the script corresponding to the task)
+		// We don't need to collect any of the outputs or execution if the task didn't execute.
+		if taskExecutionSummary != nil {
+			taskSummary.ExpandedOutputs = taskHashTracker.GetExpandedOutputs(taskSummary.TaskID)
+			taskSummary.Execution = taskExecutionSummary
+			mu.Lock()
+			taskSummaries = append(taskSummaries, taskSummary)
+			// don't hold the lock while we run ec.exec
+			mu.Unlock()
```

Comment:
I'm not sure if this comment applies anymore. Would the `defer mu.Unlock()` be appropriate if I append after `ec.exec`?  Or is it still not safe? cc @gsoltis

Reply:
As structured now, the comment is no longer relevant, since a `defer my.Unlock()` immediately after `mu.Lock()` would not hold a lock while `ec.exec` is running. It's a judgement call as to whether to switch to a `defer` here. On the one hand, it is currently trivial to see that it wouldn't hold the lock for anything critical, on the other hand, `defer` is function-scoped, not block-scoped, so it's slightly weird to have it in an `if` block.

Personally I probably wouldn't switch to defer here due to the if block, but it's no longer a huge deal since the whole thing is after `ec.exec`.

Reply:
done. I was a little unsure how `defer` would work in an anonymous function, but it works as expected https://play.golang.com/p/aEodOfXGPtV